File: 0426-convert-binary-search-tree-to-sorted-doubly-linked-list/0426-convert-binary-search-tree-to-sorted-doubly-linked-list.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;

    Node() {}

    Node(int _val) {
        val = _val;
        left = NULL;
        right = NULL;
    }

    Node(int _val, Node* _left, Node* _right) {
        val = _val;
        left = _left;
        right = _right;
    }
};
*/

class Solution {
public:
    Node* last = nullptr;
    Node* first = nullptr;
    
    Node* treeToDoublyList(Node* root) {
        if(root == nullptr) return nullptr;
        
        help(root);
        
        first->left = last;
        last->right = first;
        
        return first;
    }
    
    void help(Node* root)
    {
        if(root == nullptr) return;
        
        help(root->left);
        
        if(last == nullptr) first = root;
        
        if(last != nullptr)
        {
            last->right = root;
            root->left = last;
        }
        
        last = root;
        
        help(root->right);
        
    }
};

File: 0426-convert-binary-search-tree-to-sorted-doubly-linked-list/0426-convert-binary-search-tree-to-sorted-doubly-linked-list.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
"""

class Solution:
    def treeToDoublyList(self, root: 'Optional[Node]') -> 'Optional[Node]':
        if not root:
            return root
        
        first = None
        last = None
        
        def dfs(node):
            nonlocal first, last
            if not node:
                return
            
            
            dfs(node.left)
            
            if last == None:
                first = node
            else:
                last.right = node
                node.left = last
            
            last = node
            
            dfs(node.right)
            
        dfs(root)
        
        first.left = last
        last.right = first
        
        return first